File: JavaScript/petooh.js
/* globals define */
/* jshint node: true */

(function (root, factory) {
    'use strict';

    if (typeof define === 'function' && define.amd) {
        define([], factory);
    }
    else if (typeof exports === 'object') {
        module.exports = factory();
    }
    else {
        root.Petooh = factory();
  }
}(this, function () {
    'use strict';

    // constructor desu
    var Petooh = function (options) {
        this.options = options || {};
        this.stack = {};
        this.level = 0;
        this.currentPosition = 0;
        this.word = '';
        this.filter = new RegExp(/^[adehkKoOru]$/);
        this.brain = [];
        this.result = '';
        this.buffer = '';
    };

    // ears of your rooster
    Petooh.prototype.listen = function (error, sound) {
        if (error) {
            this.peckError();
        }

        // can you do it a bit deeper?
        if (sound.length > 1) {
            var self = this;

            // great.. no more IE 8- here..
            sound.split('').forEach(function (piece) {
                self.listen(error, piece);
            });
        }

        // ignore incomprehensible sounds plz
        if (!this.filter.test(sound)) {
            return;
        }

        var buffer = this.word + sound;

        if (buffer === 'Ko') {
            this.level > 0
                ? this.remember(buffer)
                : this.increasePlz();

            this.forget();
        }
        else if (buffer === 'kO') {
            this.level > 0
                ? this.remember(buffer)
                : this.decreasePlz();

            this.forget();
        }
        else if (buffer === 'Kudah') {
            this.level > 0
                ? this.remember(buffer)
                : this.currentPosition++;

            this.forget();
        }
        else if (buffer === 'kudah') {
            this.level > 0
                ? this.remember(buffer)
                : this.currentPosition--;

            this.forget();
        }
        else if (this.word === 'Kud' && sound !== 'a') {
            this.stack[++this.level] = [];
            this.forget(sound);
        }
        else if (this.word === 'kud' && sound !== 'a') {
            this.repeat();
            this.level--;
            this.forget(sound);
        }
        else if (buffer === 'Kukarek') {
            this.level > 0
                ? this.remember(buffer)
                : this.success();

            this.forget();
        }
        else {
            this.word += sound;
        }
    };

    Petooh.prototype.repeat = function () {
        var self = this;

        while (this.brain[this.currentPosition] > 0) {
            this.stack[this.level].forEach(function (word) {
                if      (word === 'Ko')      { self.increasePlz(); }
                else if (word === 'kO')      { self.decreasePlz(); }
                else if (word === 'Kudah')   { self.currentPosition++; }
                else if (word === 'kudah')   { self.currentPosition--; }
                else if (word === 'Kukarek') { self.success(); }
            });
        }
    };

    // catch some error
    Petooh.prototype.peckError = function () {
        throw new Error('peck-peck');
    };

    Petooh.prototype.forget = function (sound) {
        this.word = sound === void 0 ? '' : sound;
    };

    Petooh.prototype.increasePlz = function () {
        this.brain[this.currentPosition] === void 0
            ? this.brain[this.currentPosition] = 1
            : this.brain[this.currentPosition]++;
    };

    Petooh.prototype.decreasePlz = function () {
        this.brain[this.currentPosition] === void 0
            ? this.brain[this.currentPosition] = 1
            : this.brain[this.currentPosition]--;
    };

    Petooh.prototype.remember = function (word) {
        if (!this.stack[this.level]) {
            this.peckError();
        }

        this.stack[this.level].push(word);
    };

    Petooh.prototype.success = function () {
        this.result += String.fromCharCode(this.brain[this.currentPosition]);

        if (this.options.cleanBrain) {
            this.brain = [];
        }
    };

    Petooh.prototype.told = function () {
        return this.result;
    };

    return Petooh;
}));

File: CSharp/Petooh.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace Petooh
{
    class Petooh
    {
        private List<string>[] stack = new List<string>[32000];
        private int level = 0;
        private int currentPosition = 0;
        private string word = string.Empty;
        private Regex filter = new Regex(@"[^adehkKoOru]");

        private int?[] brain = new int?[1 << 16 - 1];
        private string result = string.Empty;

        private Dictionary<string, Action> OpCodes { get; set; }

        public Petooh()
        {
            OpCodes = new Dictionary<string, Action>
                      {
                          {"Ko",()=>IncreasePlz()},
                          {"kO",()=>DecreasePlz()},
                          {"Kudah",()=>++currentPosition},
                          {"kudah",()=>--currentPosition},
                          {"Kukarek",()=>Success()}
                      };
        }

        // ears of your rooster
        public void Listen(string sound)
        {
            sound=filter.Replace(sound, "");
            foreach (var piece in sound.Select(x => x.ToString()))
            {
                var buffer = word + piece;

                if (word == "Kud" && piece != "a")
                {
                    stack[++level] = new List<string>();
                    Forget(piece);
                }
                else if (word == "kud" && piece != "a")
                {
                    Repeat();
                    --level;
                    Forget(piece);
                }
                else if (OpCodes.ContainsKey(buffer))
                {
                    if (level > 0)
                    {
                        Remember(buffer);
                    }
                    else
                    {
                        OpCodes[buffer]();
                    }

                    Forget(null);
                }
                else
                {
                    word += piece;
                }
            }
        }

        void Repeat()
        {
            while (brain[currentPosition] > 0)
            {
                foreach(var word in stack[level])
                {
                    if (OpCodes.ContainsKey(word))
                    {
                        OpCodes[word]();
                    }
                }
            }
        }

        // catch some error
        void PeckError()
        {
            throw new Exception("peck-peck");
        }

        void Forget(string sound)
        {
            word = sound ?? "";
        }

        void IncreasePlz()
        {
            brain[currentPosition] = (brain[currentPosition] ?? 0) + 1;
        }

        void DecreasePlz()
        {
            brain[currentPosition] = (brain[currentPosition] ?? 2) - 1;
        }

        void Remember(string word)
        {
            if (level >= stack.Length)
            {
                PeckError();
            }
            stack[level].Add(word);
        }

        void Success()
        {
            result += Convert.ToChar(brain[currentPosition]);
        }

        public string Told
        {
            get
            {
                return result;
            }
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            if (args.Length > 0)
            {
                var pet = new Petooh();

                pet.Listen(System.IO.File.ReadAllText(args[0]));
                Console.WriteLine(pet.Told);
            }
            else
            {
                Console.WriteLine("USAGE: Petooh.exe /path/to/file.koko");
            }
            Console.WriteLine();
            Console.WriteLine("Press ANY key");
            Console.ReadKey();
        }
    }
}
